usr/share/rear/lib/authtoken-functions.sh

```diff
@@ -0,0 +1,546 @@
+#!/bin/bash
+
+function authtkn_load() {
+	# $1 container path
+	# $2 container offset
+	# $3 encryption key
+	# $4 [bound_to_blockdev path]
+
+	local intermediate_key=""
+	local cipher
+	local plain
+	local ret
+
+	cipher=$(token_read "$1" "$2"); ret=$?
+	if [ $ret -ne 0 ]; then
+		echo "authtkn_load(): token_read() failed: $ret" >&2
+		return $ret
+	fi
+
+	if [ -n "$4" ]; then
+		intermediate_key=$(blkdev_hash "$4" 2); ret=$?
+		if [ $ret -ne 0 ]; then
+			echo "authtkn_load(): failed to derive intermediate key from $4: $ret" >&2
+			return $ret
+		fi
+	fi
+
+	plain=$(decrypt_base64 "$cipher" "$3" "$intermediate_key"); ret=$?
+	if [ $ret -ne 0 ]; then
+		echo "authtkn_load(): decrypt_base64() failed: $ret" >&2
+		return $ret
+	fi
+
+	echo $plain
+	return 0
+}
+
+function authtkn_store() {
+	# $1 container path
+	# $2 container offset
+	# $3 token plaintext
+	# $4 encryption key
+	# $5 [bind_to_blockdev path]
+
+	local intermediate_key=""
+	local cipher
+	local tknlen
+	local ret
+
+	if [ -n "$5" ]; then
+		intermediate_key=$(blkdev_hash "$5" 2); ret=$?
+		if [ $ret -ne 0 ]; then
+			echo "authtkn_store(): failed to derive intermediate key from $5: $ret" >&2
+			return $ret
+		fi
+	fi
+
+	cipher=$(encrypt_base64 "$3" "$4" "$intermediate_key"); ret=$?
+	if [ $ret -ne 0 ]; then
+		echo "authtkn_store(): encrypt_base64() failed: $ret" >&2
+		return $ret
+	fi
+
+	tknlen=$(token_write "$cipher" "$1" "$2"); ret=$?
+	if [ $ret -ne 0 ]; then
+		echo "authtkn_store(): token_write() failed: $ret" >&2
+		return $ret
+	fi
+
+	echo $tknlen
+	return 0
+}
+
+function authtkn_wipe() {
+	# $1 container path
+	# $2 container offset
+	# $3 [token length]
+
+	# ensure byte-length counting
+	LANG=C LC_ALL=C
+
+	local ret
+	local length
+	local token
+
+	if [ -z "$3" ]; then
+		# taking existing token length
+		token=$(token_read "$1" "$2")
+		[[ $? -eq 0 ]] && length=${#token}
+	else
+		length=$3
+	fi
+
+	# will fail if length still undef
+	token=$(garbage_base64 $length); ret=$?
+	if [ $ret -ne 0 ]; then
+		echo "authtkn_wipe(): garbage_base64() failed: $ret" >&2
+		return $ret
+	fi
+
+	length=$(token_write "$token" "$1" "$2"); ret=$?
+	if [ $ret -ne 0 ]; then
+		echo "authtkn_wipe(): token_write() failed: $ret" >&2
+		return $ret
+	fi
+
+	echo $length
+	return 0
+}
+
+
+function token_read() {
+	# $1 path
+	# $2 [offset:0]
+	# $3 [retries:2]
+
+	# ensure byte-length counting
+	LANG=C LC_ALL=C
+
+	if [ -z "$1" ]; then
+		echo "token_read(): none or empty token container path given" >&2
+		return 1
+	fi
+
+	local rawchunk
+    local length
```

Comment:
Don't use tabs for indentation because too easily
it gets messy when tabs and spaces are used
intermixed for indentation, see
https://github.com/rear/rear/wiki/Coding-Style#text-layout